Initial ConfigurationBuilder API changes

HaoK opened the pull request:
# Overview:

Covers most of what we had discussed earlier this week.

Lou had an idea to split the configuration of the providers into sources, and build instances later.  This now eliminates the point in time behavior of the builders (FileProviders are only locked when the configurationBuilder is Build()).

There's now a base abstract FileConfigurationSource/Provider which makes the implementations of Json/Ini/Xml pretty lean as all the boilerplate has been lifted including the extension goop.  

## Usage patterns:

Simple common usage is unaffected:
```
    new ConfigurationBuilder().AddJson("foo.json").AddXml("bar.xml").Build();
```

Setting only the base path:
```
    new ConfigurationBuilder()
      .SetBasePath(@"C:\SomeOtherDirectory")
      .AddJson("foo.json").AddXml("bar.xml").Build();
```

Setting optional or reload on change:
```
    new ConfigurationBuilder()
      .AddJson(source => {
          source.Path = "foo.json";
          source.Optional = true;
       }).AddXml(source => {
          source.Path = "bar.xml";
          source.ReloadOnChanged = false;
       }).Build();
```

Full explosion of setting everything
```
    new ConfigurationBuilder()
      .SetBasePath(@"C:\SomeOherDirectory")
      .AddJson(source => {
          source.Path = "foo.json";
          source.Optional = true;
          source.ReloadOnChanged = false;
       }).AddXml(source => {
          source.Path = "bar.xml";
          source.Optional = true;
          source.ReloadOnChanged = false;
       }).Build();
```



I haven't finished wiring up the reload token stuff, I wanted to get the initial PR out for feedback on surface area, existing tests are passing, but there isn't a lot of tests on the new surface area/behavior yet

 @divega @shirhatti @lodejard @davidfowl

lodejard commented:
Looks like a great improvement to me so far. @eilon had the additional note that we should have at a minimum the same extension methods we did before, so that will call for an "optional" bool overload I guess.

Also, I have to reiterate the recommendation I made before. Don't put the optional and reloadable bools into the properties bucket and remove the ConfigureFileSourceDefaults method. The extra surface area isn't justified by what they provide.

HaoK commented:
I guess I was pushing for the defaults to try and move away entirely from the overloads, basically there's one `Add(path)` which using all the default settings, and one overload that takes a delegate. 

So these 3?
```
  AddXyz(string path);
  AddXyz(string path, bool optional);
  AddXyz(Action<XyzSource> configureSource);
```

MichaCo commented:
Just a cosmetic thing but, personally I don't like arrow functions which have to have method bodies, especially in a builder.

So you have this now

    AddJson(source => {
          source.Path = "foo.json";
          source.Optional = true;
          source.ReloadOnChanged = false;
       }

which is nice, but why not make the `source` in this case a builder, too, like 

    AddJson(source => source
          .WithPath("foo.json")
          .IsOptional(true)
          .WithReloadOnChanged(false));

divega commented on src/Microsoft.Extensions.Configuration.FileExtensions/FileConfigurationProvider.cs:
```diff
@@ -0,0 +1,55 @@
+        /// <summary>
+        /// Gets a value that determines if this instance of <see cref="FileConfigurationProvider"/> is optional.
+        /// </summary>
+        public bool Optional { get; set; }
```
Do we really want these properties to be settable all the time or is this just an artifact of the implementation? It seems that it was cleaner in the previous code.

HaoK replied:
You can't actually get a hold of the provider instances anymore, they aren't exposed off of IConfiguration and they are directly fed to ConfigurationRoot via the Build method.

HaoK replied:
You can think of the providers are implementation worker objects (they are effectively private)...  The Source objects are the thing that you tweak/configure

divega replied:
I mean "cleaner" so I am talking perspective of the quality of the codebase. I would just make each concrete `FileConfigurationSource` source implementation pass these in the constructor and I would then make properties that don't make sense to change read-only again.

HaoK replied:
That doesn't work as well with how I implemented things currently via `AddFileSource<TSource>` which new's up the actual TSource and sets the base properties.  That method currently has a new() constraint on the source.

HaoK replied:
in regards to 'cleaner'.  All things being equal I think a POCO with no ctor and only get/set properties is cleaner anyways.

HaoK replied:
As once you have a ctor, all the derived classes have to have it and call base feeding in parameters..that is way more boilerplate

divega replied:
The source is a POCO with a default constructor that is only used with the builder, and that is fine. I am talking about the provider, in case it isn't clear.

lodejard replied:
The provider doesn't really need to be public when it comes right down to it...

Or if it should stay public, another thing to consider is having the source create a provider by using itself as a single constructor argument with `return new XxxProvider(this)`. The provider could then copy the members into protected or get only properties.

divega replied:
I prefer to leave them public for people to be able to extend from them. :+1: on passing the source in the constructor and coping the values to get-only properties.

HaoK commented:
@MichaCo Perhaps this is a taste thing, but I find the sentence like fluent to be pretty'wordy'/verbose looking...  As it currently stands it more or less identical to how Options are configured in startup via setting poco properties...

MichaCo commented:
@HaoK 
> this is a taste thing

totally true :>

divega commented:
>> this is a taste thing

>totally true :>

:+1: And I personally prefer the fluent style @MichaCo proposed but for something this small I don't think it matters too much.

divega commented:
@lodejard said
> Looks like a great improvement to me so far. @eilon had the additional note that we should have at a minimum the same extension methods we did before, 

Personally I am still not sure it was worth it. Given we end up with a superset of the surface this was strictly an internal refactoring that resulted in the addition of a new concept (the new sources) and a new interface that providers need to implement. IMO the way this was implemented in the AddXyz methods and the provider constructors wasn't that bad even if there was some duplication.

> so that will call for an "optional" bool overload I guess.

Isn't `reloadOnChange` important enough to have a first class overload? If it isn't then what principle are we using to decide?

> Also, I have to reiterate the recommendation I made before. Don't put the optional and reloadable bools into the properties bucket and remove the ConfigureFileSourceDefaults method. The extra surface area isn't justified by what they provide.

:+1:

lodejard commented on src/Microsoft.Extensions.Configuration.Abstractions/IConfigurationSource.cs:
```diff
@@ -0,0 +1,20 @@
+        /// </summary>
+        /// <param name="builder">The <see cref="IConfigurationBuilder"/>.</param>
+        /// <returns></returns>
+        IConfigurationProvider Build(IConfigurationBuilder builder);
```
Only needs IDictionary properties, is there a reason to pass builder?

HaoK replied:
If we pass the builder we can use the extension methods to get the FileProvider, otherwise we have to hardcode constants for the keys

lodejard replied:
:ok: I'll nuke the note

lodejard commented on src/Microsoft.Extensions.Configuration.FileExtensions/FileConfigurationExtensions.cs:
```diff
@@ -2,23 +2,90 @@
+                throw new ArgumentNullException(nameof(fileProvider));
+            }
+
+            builder.GetFileSourceDefaults().FileProvider = fileProvider;
```
Should be `builder.Properties["FileProvider"] = fileProvider`

lodejard commented on src/Microsoft.Extensions.Configuration.FileExtensions/FileConfigurationExtensions.cs:
```diff
@@ -2,23 +2,90 @@
+        /// <param name="builder">The <see cref="IConfigurationBuilder"/> to add to.</param>
+        /// <param name="configureDefaults">Configures the default <see cref="FileConfigurationSource"/>.</param>
+        /// <returns>The <see cref="IConfigurationBuilder"/>.</returns>
+        public static IConfigurationBuilder ConfigureFileSourceDefaults(this IConfigurationBuilder builder, Action<FileConfigurationSource> configureDefaults)
```
Remove this - it's extra api for something you can already do.

lodejard commented on src/Microsoft.Extensions.Configuration.FileExtensions/FileConfigurationExtensions.cs:
```diff
@@ -2,23 +2,90 @@
+            return builder;
+        }
+
+        public static FileConfigurationSource GetFileSourceDefaults(this IConfigurationBuilder configurationBuilder)
```
Remove this too

lodejard commented on src/Microsoft.Extensions.Configuration.FileExtensions/FileConfigurationExtensions.cs:
```diff
@@ -2,23 +2,90 @@
         /// <param name="basePath">The absolute path of file-based providers.</param>
         /// <returns>The <see cref="IConfigurationBuilder"/>.</returns>
-        public static IConfigurationBuilder SetBasePath(this IConfigurationBuilder configurationBuilder, string basePath)
+        public static IConfigurationBuilder SetFileProvider(this IConfigurationBuilder builder, string basePath)
```
This should be named `.SetBasePath(basePath)`

lodejard commented on src/Microsoft.Extensions.Configuration.FileExtensions/FileConfigurationExtensions.cs:
```diff
@@ -26,38 +93,33 @@ public static IConfigurationBuilder SetBasePath(this IConfigurationBuilder confi
+        /// <param name="builder">The <see cref="IConfigurationBuilder"/> to add to.</param>
+        /// <param name="configureSource">Configures the <see cref="IConfigurationSource"/> to add.</param>
+        /// <returns>The <see cref="IConfigurationBuilder"/>.</returns>
+        public static IConfigurationBuilder AddFileSource<TSource>(this IConfigurationBuilder builder, Action<TSource> configureSource)
```
Remove this - it's just another way of doing what the extension methods do, but it's less discoverable because you need to know the TSource in advance.

lodejard commented on src/Microsoft.Extensions.Configuration.FileExtensions/FileConfigurationSource.cs:
```diff
@@ -0,0 +1,99 @@
+    /// <typeparam name="TProvider">The concrete FileConfigurationProvider to create.</typeparam>
+    public class FileConfigurationSource<TProvider> : FileConfigurationSource where TProvider : FileConfigurationProvider, new()
+    {
+        public override IConfigurationProvider Build(IConfigurationBuilder builder)
```
Simpler just to implement Build in the concrete source classes by calling new on the concrete provider classes. You don't need this generic layer at all.

lodejard commented on src/Microsoft.Extensions.Configuration.FileExtensions/FileConfigurationSource.cs:
```diff
@@ -0,0 +1,99 @@
+        /// </summary>
+        public bool ReloadOnChange { get; set; }
+
+        public virtual IFileInfo GetFileInfo()
```
Source should have data, not behavior. It shouldn't have a GetFileInfo method.

lodejard commented on src/Microsoft.Extensions.Configuration.FileExtensions/FileConfigurationSource.cs:
```diff
@@ -0,0 +1,99 @@
+        /// Applies the defaults specified from the builder.
+        /// </summary>
+        /// <param name="builder"></param>
+        public virtual void ApplyFileSourceDefaults(IConfigurationBuilder builder)
```
Remove method - there's no such thing as file source defaults. Just the FileProvider property.

lodejard commented on src/Microsoft.Extensions.Configuration.FileExtensions/FileConfigurationSource.cs:
```diff
@@ -0,0 +1,99 @@
+        public virtual void InitializeProvider(IConfigurationBuilder builder, FileConfigurationProvider provider)
+        {
+            ApplyFileSourceDefaults(builder);
+            provider.File = GetFileInfo();
```
Provider should have IFileProvider and path properties, not a File property. You shouldn't close over and keep alive the same IFileInfo.

lodejard commented on src/Microsoft.Extensions.Configuration.FileExtensions/FileConfigurationSource.cs:
```diff
@@ -0,0 +1,99 @@
+            // TODO: aggregate watch?
+            if (ReloadOnChange && FileProvider != null)
+            {
+                ChangeToken.OnChange(() => FileProvider.Watch(Path), () => provider.Load());
```
This ChangeToken pump should be in the provider, not the source. The source's only job is to hold onto properties, and call new on the provider.

lodejard commented on src/Microsoft.Extensions.Configuration.Json/JsonConfigurationExtensions.cs:
```diff
@@ -57,16 +30,18 @@ public static class JsonConfigurationExtensions
+        /// <returns>The <see cref="IConfigurationBuilder"/>.</returns>
+        public static IConfigurationBuilder AddJsonFile(this IConfigurationBuilder builder, Action<JsonConfigurationSource> configureSource)
+        {
+            return builder.AddFileSource(configureSource);
```
```cs
  var source = new JsonConfigurationSource();
  configureSource(source);
  return builder.Add(source);
```

Same pattern each place. Let's not use generic magic when simple code works.

lodejard commented on src/Microsoft.Extensions.Configuration.Xml/XmlConfigurationProvider.cs:
```diff
@@ -10,84 +10,15 @@
         private const string NameAttributeKey = "Name";
 
-        private readonly XmlDocumentDecryptor _xmlDocumentDecryptor;
+        internal XmlDocumentDecryptor Decryptor { get; set; } = XmlDocumentDecryptor.Instance;
```
`internal` keyword is an anti-pattern. Should this be private or public?

HaoK replied:
XmlDocumentDecryptor appears to be an internal class which looks only to be used for testing, hence the property is internal.  I guess we can make all of these things public if you prefer...

lodejard replied:
Yeah, internal for testing is best avoided.

HaoK replied:
Ok I'll make XmlDocumentDecryptor and the property public

lodejard commented on src/Microsoft.Extensions.Configuration/ConfigurationBuilder.cs:
```diff
@@ -30,41 +31,37 @@ public IEnumerable<IConfigurationProvider> Providers
+            foreach (var source in _sources)
+            {
+                var provider = source.Build(this);
+                // REVIEW: who should call load?  source.Build?
```
"who should call load" - probably the provider itself, at the same place that hooks up the change token reloader if it has been so configured.

HaoK replied:
So the ctor of the provider?

HaoK replied:
I think the builder should actually call Load, I tried it via the provider but it just seems like that makes it easy to forget to call Load and have weird behavior.

Feels better to just have the Builder do it before Add()...

```
                var provider = source.Build(this);
                provider.Load();
                providers.Add(provider);
```

lodejard commented on src/Microsoft.Extensions.Configuration/DefaultConfigurationBuilderExtensions.cs:
```diff
@@ -21,7 +21,7 @@ public static IConfigurationBuilder Include(this IConfigurationBuilder configura
                 throw new ArgumentNullException(nameof(configurationBuilder));
             }
-            configurationBuilder.Add(new IncludedConfigurationProvider(includedConfiguration));
+            configurationBuilder.Add(new IncludedConfigurationSource { Source = includedConfiguration });
```
We should cut chaining configuration altogether. It was only needed because of Hosting.dll, which no longer requires it.

HaoK replied:
@divega nuke Include config?

divega replied:
It was a cool feature, but we didn't think the current implementation was ideal (remember we have a bug to make avoid copying the state?). Also we would need to wire up change notifications to make it complete.

So if it is confirmed that nobody is using it, I agree removing it is the best option.

cc @davidfowl @DamianEdwards who originally insisted it was necessary, just in case.

lodejard commented on src/Microsoft.Extensions.Configuration/MemoryConfigurationSource.cs:
```diff
@@ -0,0 +1,50 @@
+
+        private class MemoryConfigurationProvider :
+            ConfigurationProvider,
+            IEnumerable<KeyValuePair<string, string>>
```
The `IEnumerable<KeyValuePair<string, string>>` and `Add(key,value)` are relocated to the MemoryConfigurationSource

lodejard commented:
Yeah, let's really focus this change on: Supporting IFileProvider, Adding reloadOnChange. The introduction of the ConfigurationSource layer is really about making those two things possible. Beyond that - let's not dive into an API revamp unless we file a seperate issue specifically intended to revamp API. Which I can almost guarantee wouldn't get a thumbs up at this point. :)

divega commented on src/Microsoft.Extensions.Configuration.Abstractions/IConfigurationSource.cs:
```diff
@@ -0,0 +1,20 @@
+namespace Microsoft.Extensions.Configuration
+{
+    /// <summary>
+    /// Represents a source of configuration key/values for an application.
```
Isn't this exactly the description we had on `IConfigurationProvider`? Let's try to come up with a descriptions that are clear and more distinguishable. E.g. this type only represents a source of configuration in the builder.

divega commented:
I agree with @lodejard's suggestions in this PR so far. If these are taken I can see more clearly the value of this change.

HaoK commented:
Updated: nuked default logic, Providers take the Source in their ctors and use that as their settings now instead of having properties

HaoK commented:
So open question, we had several tests that were verifying functionality directly against the ConfigurationProvider instances before.  This no longer is possible with the new surface area since we can't get to the ConfigurationProvider instances anymore.  What do we want to do with these tests?

They were verifying things like SettingValueUpdatesAllConfigurationProviders

HaoK commented:
@lodejard how did you envision the reload working with both Reload on the root and each provider? 

Maybe for simplification each Provider can calling `configRoot.Reload()` instead of just reloading itself.  That way we don't have to add any complexity around managing state, if any provider trips, we reload all providers.

Otherwise we'd have pass around tokens to enable a way for the providers to get back to the ConfigurationRoot's token...

HaoK commented:
So looks like one good reason to have the providers be public is for testing, I updated the tests to recombine the source/providers into one so the providers can be examined by the tests:

```
        public class TestMemorySourceProvider : MemoryConfigurationProvider, IConfigurationSource
        {
            public TestMemorySourceProvider(Dictionary<string, string> initialData) 
                : base(new MemoryConfigurationSource { InitialData = initialData })
            { }

            public IConfigurationProvider Build(IConfigurationBuilder builder)
            {
                return this;
            }
        }
```

HaoK commented:
Updated to reenable tests and wire up provider reloads triggering the reload token on ConfigurationRoot.

Needed to introduce some more plumbing:
```
// IConfigurationRoot methods         
        void RaiseChanged();
        IDisposable RegisterOnChanged(Action<object> callback, object state);

// IConfigurationProvider needs the root to call RaiseChanged as part of file watching
        /// <summary>
        ///  Called by the root after it has finished being built so providers can do any initialization.
        /// </summary>
        /// <param name="root"></param>
        void Initialize(IConfigurationRoot root);

```

lodejard commented on src/Microsoft.Extensions.Configuration.FileExtensions/FileConfigurationProvider.cs:
```diff
@@ -0,0 +1,80 @@
+                    () =>
+                    {
+                        Load();
+                        root.RaiseChanged();
```
Would it also work to have IConfigurationProvider expose an individual reload token, and have the root's reload token expire when any provider's reload token is expired? That way the IConfiguration wouldn't need extra APIs like RaiseChanged.

HaoK replied:
That would work but seems a bit complicated.  Maybe we should just introduce some kind of IChangeTracker to encapsulate the ability to fire changes more than once, and pass the root's tracker to the providers.  We just need something that's more persistent since the change token only works for one firing today...

We could do it scoped to only config or add it primitives and make it general purpose.

```
   public interface IConfigurationChangeTracker {
      public void RegisterOnChanged(Action<IConfigurationRoot> callback)
      public void RaiseChanged();
   }

   IConfigurationRoot would add a ChangeTracker readonly property
   
```

HaoK commented:
Updated with final tweaks, added IChangeToken to IConfigurationProvider with a base implementation.  @divega can you do a final review?

I'll submit a new PR for the reload changes as a stand alone followup change

HaoK commented:
30e59f1190e01a9ae0c2fae4022af5ba3ca2eafd

poke commented:
From [a comment above](https://github.com/aspnet/Configuration/pull/407#issuecomment-195534813):
>  As it currently stands it more or less identical to how Options are configured in startup via setting poco properties...

Out of curiosity: Is there a particular reason why this lambda function pattern is being used all over the place instead of passing POCOs?

```
cb.AddJson(source => {
    source.Path = "foo.json";
    source.Optional = true;
});
```

This does not seem that much of a difference compared to the following:

```
cb.AddJson(new JsonConfigurationSource {
    Path = "foo.json",
    Optional = true
});
```

Except that the overhead would likely be lower in this case.

divega commented:
@poke the main advantage of the lambda for fluent APIs like this is that you don't need to know or write the type nor its namespace. The instance can also be pre-populated with default values that are method specific.

poke commented:
@divega That makes sense, thank you for the explanations!

HaoK commented:
You can already pass a source directly via IConfigurationBuilders Add method if that is what you desire.

PinpointTownes commented:
@divega too bad the security middleware no longer leverages this "fluent APIs/delegate extensions" style... I really miss it: https://github.com/aspnet/Options/issues/105#issuecomment-168866368.
